Introduce new CRD for EgressGW

julianwiedmann opened the pull request:
As part of promoting the EgressGW feature to stable, introduce a new CRD type and deprecate the old one.  The new CRD allows for better selection of the Egress Node & IP.

```release-note
Introduce a new CRD (CiliumEgressGatewayPolicy) for Egress Gateway configuration. Deprecate the previous CRD (CiliumEgressNATPolicy).
```

julianwiedmann committed:
egressgw: add new CiliumEgressGatewayPolicy CRD

As part of graduating the EgressGW feature from beta status, introduce a
fresh CiliumEgressGatewayPolicy CRD that only supports the egressGateway
property for configuring an EgressGW. Also rename one of the existing
fields for clarity ('egress' vs 'selectors').

Then remove the recently introduced support for egressGateway from the CENP
CRD. We rather want to phase this CRD out.

For testing, we have
- basic tests for CEGP
- a legacy test for CENP
- an "upgrade" test with the same policy applied via CENP and CEGP,
  and the CENP being removed.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

julianwiedmann committed:
egressgw: deprecate the CENP CRD

Warn users that the CENP-style of configuring an EgressGW is going away
in v1.13, and encourage them to switch to CEGP instead.

Also let the agent log any usage of CENP resources.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

jibi commented on test/k8s/manifests/egress-nat-policy.yaml:
```diff
@@ -63,12 +63,12 @@ spec:
 ---
-apiVersion: cilium.io/v2alpha1
-kind: CiliumEgressNATPolicy
+apiVersion: cilium.io/v2
```
This should stay `v2alpha1` as it's still using the old `egressSourceIP` field (beside that tests seem happy).
We can then remove it once we get rid of the old API

jibi commented on pkg/egressgateway/policy.go:
```diff
@@ -299,3 +286,94 @@ func ParsePolicyConfigID(cenp *v2alpha1.CiliumEgressNATPolicy) types.NamespacedN
+
+// ParseCEGP takes a CiliumEgressGatewayPolicy CR and converts to PolicyConfig,
+// the internal representation of the egress gateway policy
+func ParseCEGP(cegp *v2.CiliumEgressGatewayPolicy) (*PolicyConfig, error) {
```
nit: maybe rename also `ParsePolicy` to `ParseCENP`? (Or even better if possible refactor the 2 into a single function that takes the API version as parameter)

julianwiedmann replied:
Rename sounds good! Will have another look at consolidating the two functions, but maybe rather as a follow-up.

julianwiedmann committed:
docs: document NO_CILIUM_ON_NODE[s]

Make it a bit easier to spot this parameter.

While at it also fix a minor formatting bug for SERVER_BOX.

Signed-off-by: Julian Wiedmann <jwi@isovalent.com>

qmonnet commented on Documentation/contributing/testing/e2e.rst:
```diff
@@ -637,6 +637,8 @@ configuration options that can be passed as environment variables:
 +======================+===================+==============+==================================================================+
 | K8S\_NODES           | 2                 | 0..100       | Number of Kubernetes nodes in the cluster                        |
 +----------------------+-------------------+--------------+------------------------------------------------------------------+
+| NO_CILIUM_ON_NODE[S] | none              | *            | Comma-separated list of K8s nodes that should not run Cilium     |
```
Nit: the star is rendered as a list item bullet, you might want to escape it with a backslash.
![image](https://user-images.githubusercontent.com/17001771/166491845-78542de0-182a-419c-bc66-43b4efb0c9e4.png)

julianwiedmann replied:
Ah indeed - thanks, fixed. Also fixed up the same issue for `SERVER_BOX`.

jibi commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -13,9 +13,110 @@ import (
+// +kubebuilder:printcolumn:JSONPath=".metadata.creationTimestamp",name="Age",type=date
+// +kubebuilder:storageversion
+
+type CiliumEgressGatewayPolicy struct {
```
This should probably live in its own file (`cegp_types.go`)

julianwiedmann replied:
Can do, was just sticking to the `v2alpha1` layout.

jibi commented on test/k8s/egress.go:
```diff
@@ -264,6 +264,37 @@ var _ = SkipDescribeIf(func() bool {
 				})
 			})
+
+			Context("egress gw policy upgrade", func() {
```
Nice :100:

jibi commented on pkg/k8s/apis/cilium.io/v2alpha1/types.go:
```diff
@@ -14,9 +14,10 @@ import (
+// +kubebuilder:resource:categories={cilium,ciliumpolicy},singular="ciliumegressnatpolicy",path="ciliumegressnatpolicies",scope="Cluster",shortName={cenp}
 // +kubebuilder:printcolumn:JSONPath=".metadata.creationTimestamp",name="Age",type=date
 // +kubebuilder:storageversion
+// +kubebuilder:deprecatedversion:warning="CiliumEgressNATPolicy is deprecated and will be removed in version 1.13. Use CiliumEgressGatewayPolicy instead."
```
TIL `kubebuilder:deprecatedversion`

jibi reviewed (approved):
LGTM :ship: :ship:

ldelossa reviewed (approved):
This all LGTM. Reading this review made it dawn on me that I did not add anything about watchers.go to the CRD documentation I submitted a bit earlier. If you'd like you can also update those user facing docs with instructions on how to plug into watcher's infrastructure.
